File: src/24-challenge/solution.js
/**
 * @typedef {Object} BTreeNode
 * @property {string} value
 * @property {BTreeNode | undefined} left
 * @property {BTreeNode | undefined} right
 */

/**
 * @param {BTreeNode | undefined} tree01 - The first binary tree.
 * @param {BTreeNode | undefined} tree02 - The second binary tree.
 * @returns {[boolean, string]}
 */
export default function fn(tree01, tree02) {
	if (!tree01 && !tree02) return [true, '']
	if (!tree01 || !tree02) return [false, tree01?.value ?? '']

	if (tree01.value !== tree02.value) return [false, tree01.value]

	const isLeftSynchronized = fn(tree01.left, tree02.right)[0]
	const isRightSynchronized = fn(tree01.right, tree02.left)[0]

	return [isLeftSynchronized && isRightSynchronized, tree01.value]
}

File: src/24-challenge/solution.py
from typing import TypedDict, Optional


BTreeNode = TypedDict(
    "BTreeNode",
    {"value": str, "left": Optional["BTreeNode"], "right": Optional["BTreeNode"]},
)


def fn(
    tree_01: Optional[BTreeNode] = None, tree_02: Optional[BTreeNode] = None
) -> list[bool | str]:
    if not tree_01 and not tree_02:
        return [True, ""]

    if not tree_01 or not tree_02:
        return [False, tree_01["value"] if (tree_01) else ""]

    if tree_01["value"] != tree_02["value"]:
        return [False, tree_01["value"]]

    is_left_synchronized = fn(tree_01["left"], tree_02["right"])[0]
    is_right_synchronized = fn(tree_01["right"], tree_02["left"])[0]

    return [is_left_synchronized and is_right_synchronized, tree_01["value"]]
